app/packages/core/src/utils/hooks/useQueryState.tsx

```diff
@@ -0,0 +1,64 @@
+import queryString from 'query-string';
+import { useMemo, useRef } from 'react';
+import { useLocation, useNavigate } from 'react-router-dom';
+
+import useMemoizedFn from './useMemoizedFn';
+import useUpdate from './useUpdate';
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type QueryState = Record<string, any>;
+
+/**
+ * `useQueryState` is a React hook, which allows us to store the state into a url query.
+ *
+ * Note: The implementation is heavily inspired by https://ahooks.js.org/hooks/use-url-state, but only supports React
+ * Router v6, uses `react-router-dom` instead of `react-router` and removes the options, which should always be the same
+ * across all our components.
+ */
+const useQueryState = <S extends QueryState = QueryState>(initialState?: S | (() => S)) => {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  type State = Partial<{ [key in keyof S]: any }>;
+
+  const location = useLocation();
+  const navigate = useNavigate();
+  const update = useUpdate();
+
+  const initialStateRef = useRef(typeof initialState === 'function' ? (initialState as () => S)() : initialState || {});
+
+  const queryFromUrl = useMemo(() => {
+    return queryString.parse(location.search, { parseBooleans: false, parseNumbers: false });
```

Comment:
Can we use react-router-dom's [`useSearchParams`](https://reactrouter.com/en/main/hooks/use-search-params)

It returns the `URLSearchParams` object and also supports encoding the params (I already use it here for the params of the Plugins page: https://github.com/kobsio/kobs/pull/466/files#diff-37292e6db335aa81f08a79d00596e68924889e1b89a2abbdae776302a925e72cR49-R61)

Reply:
No, because this supports more advanced use cases.